File: frontend/src/api/shutdown.ts
import { apiFetch } from "./client";

export interface RemoteDevice {
  id: number;
  name: string;
  host: string;
  port: number;
  username: string;
  enabled: boolean;
  use_sudo: boolean;
  mains_state: "online" | "on_battery" | "any";
  battery_state: "charging" | "discharging" | "full" | "any";
  battery_threshold: number;
  custom_command: string | null;
  host_key_fingerprint: string | null;
  last_test_at: string | null;
  last_result: string | null;
}

export type DeviceInput = Omit<
  RemoteDevice,
  "id" | "host_key_fingerprint" | "last_test_at" | "last_result"
>;

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await apiFetch(path, init);
  if (!response.ok) {
    const body = (await response.json().catch(() => ({}))) as {
      detail?: string;
    };
    throw new Error(body.detail || `Request failed (${response.status})`);
  }
  return response.status === 204
    ? (undefined as T)
    : (response.json() as Promise<T>);
}

export const shutdownApi = {
  devices: () => request<RemoteDevice[]>("/api/shutdown/devices"),
  create: (body: DeviceInput) =>
    request<RemoteDevice>("/api/shutdown/devices", {
      method: "POST",
      body: JSON.stringify(body),
    }),
  remove: (id: number) =>
    request<void>(`/api/shutdown/devices/${id}`, { method: "DELETE" }),
  publicKey: () => request<{ public_key: string }>("/api/shutdown/public-key"),
  inspectKey: (id: number) =>
    request<{ fingerprint: string; algorithm: string }>(
      `/api/shutdown/devices/${id}/host-key`,
    ),
  trustKey: (id: number) =>
    request<{ fingerprint: string }>(
      `/api/shutdown/devices/${id}/trust-host-key`,
      {
        method: "POST",
      },
    ),
  test: (id: number) =>
    request<{ success: boolean; message: string }>(
      `/api/shutdown/devices/${id}/test`,
      {
        method: "POST",
      },
    ),
  settings: () =>
    request<{ enabled: boolean; dry_run: boolean }>("/api/shutdown/settings"),
  updateSettings: (enabled: boolean, dry_run: boolean) =>
    request<{ enabled: boolean; dry_run: boolean }>("/api/shutdown/settings", {
      method: "PUT",
      body: JSON.stringify({ enabled, dry_run }),
    }),
  simulate: (
    mains_state: string,
    battery_state: string,
    battery_percentage: number,
  ) =>
    request<
      { device_id: number; name: string; matches: boolean; reason: string }[]
    >("/api/shutdown/simulate", {
      method: "POST",
      body: JSON.stringify({ mains_state, battery_state, battery_percentage }),
    }),
};

File: backend/src/wattsup/schemas/shutdown.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

MainsState = Literal["online", "on_battery", "any"]
BatteryState = Literal["charging", "discharging", "full", "any"]


class DeviceInput(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    host: str = Field(min_length=1, max_length=255)
    port: int = Field(default=22, ge=1, le=65535)
    username: str = Field(min_length=1, max_length=100)
    enabled: bool = False
    use_sudo: bool = True
    mains_state: MainsState = "on_battery"
    battery_state: BatteryState = "discharging"
    battery_threshold: int = Field(default=30, ge=0, le=100)
    custom_command: str | None = None


class DeviceOutput(DeviceInput):
    model_config = ConfigDict(from_attributes=True)

    id: int
    host_key_fingerprint: str | None
    last_test_at: datetime | None
    last_result: str | None


class HostKeyInfo(BaseModel):
    algorithm: str
    fingerprint: str
    public_key: str
    trusted: bool


class AutomationSettings(BaseModel):
    enabled: bool
    dry_run: bool


class TestResult(BaseModel):
    success: bool
    message: str


class SimulationRequest(BaseModel):
    mains_state: Literal["online", "on_battery"]
    battery_state: Literal["charging", "discharging", "full"]
    battery_percentage: int = Field(ge=0, le=100)


class SimulationResult(BaseModel):
    device_id: int
    name: str
    matches: bool
    reason: str
